mention: Implement notification part of @topic mention.

prakhar1144 opened the pull request:
This carries forward the work in #22855

<!-- Describe your pull request here.-->
* **First Commit**: [Review comment](https://github.com/zulip/zulip/pull/22855#discussion_r962193044)
* **Second & Third Commit**: [Comment](https://github.com/zulip/zulip/pull/22855#issuecomment-1239758013)
* **Fourth Commit**: [Review Comment](https://github.com/zulip/zulip/pull/25828#discussion_r1210158168), [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/.40topic.20mention/near/1581610)
* **Fifth Commit**: Adds a 'has_topic_wildcards' instance variable to the 'MentionData' class for the detection of - possible topic wildcards mentions.
* **Sixth Commit**: Fetch users to be notified of @topic mentions.

Note:
Meaning of '**possible** topic wildcards mention': Here, the 'has_topic_wildcards' flag is set to true when there is a possibility of @topic mention. We are not sure yet, as the message rendering takes place at a later stage in the code flow.

Fixes part of #22829



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

abhijeetbodas2001 commented on zerver/lib/mention.py:
```diff
@@ -43,6 +43,12 @@ def Q(self) -> Q:
+class MentionText:
+    text: Optional[str]
+    is_wildcard: bool
+
```
Would `TypedDict` be a simpler way to do this (especially since we don't intend to add any methods to this dataclass)?

timabbott replied:
I think a dataclass is fine fo rhtis.

abhijeetbodas2001 commented on zerver/lib/mention.py:
```diff
@@ -46,12 +47,14 @@ def Q(self) -> Q:
 class PossibleMentions:
     mention_texts: Set[str]
+    message_has_topic_wildcards: bool
     message_has_wildcards: bool
```
With ref to the other comment about `stream_wildcards`, `message_has_wildcards` here would become `message_has_stream_wildcards`, which I think reads nicely. What do you think?

prakhar1144 replied:
I added a prep commit to address this. (Fourth commit)

abhijeetbodas2001 reviewed (commented):
Thanks for rebasing this!

prakhar1144 commented on zerver/lib/mention.py:
```diff
@@ -18,6 +18,7 @@
 
+topic_wildcards = frozenset(["topic"])
 wildcards = frozenset(["all", "everyone", "stream"])
 
```
Agreed; renaming it `stream_wildcards` makes things more clear.

So, we should add a refactor commit to rename `wildcards` to `stream_wildcards` internally, keeping the `wildcard_mentioned` flag of `UserMessage` intact.

We don't plan to have a separate `UserMessage` flag for `topic_wildcards`.

I will open a topic on CZO to get more input on this, and then I'll refactor accordingly.

prakhar1144 replied:
I added a prep commit to address this. (Fourth commit)

[CZO Discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/.40topic.20mention/near/1581610)

prakhar1144 commented:
@zulipbot add "mentor review"

abhijeetbodas2001 commented:
https://chat.zulip.org/#narrow/stream/3-backend/topic/.40topic.20mention

prakhar1144 commented on zerver/tests/test_soft_deactivation.py:
```diff
@@ -632,6 +633,7 @@ def assert_num_possible_users(
+                        possibly_topic_wildcard_mentioned_user_ids=possibly_topic_wildcard_mentioned_user_ids,
                         possibly_mentioned_user_ids=possibly_mentioned_user_ids,
                     )
                 ),
```
I would be adding the following code block at line 771 of this file in later commits:

```python
 # Test UserMessage row is created while user is deactivated if there is a topic wildcard mention
 # i.e. @topic
    assert_stream_message_sent_to_idle_user(
         "Test @**topic** mention", possibly_topic_wildcard_mentioned_user_ids={long_term_idle_user.id}
    )
    assert_stream_message_not_sent_to_idle_user("Test @**bogus** mention")

```
Not adding here as `assert_stream_message_sent_to_idle_user` (line 643) tests things other than `len(get_subscriptions_for_send_message())` which will be added in later commits, so tests can be added at that time.

prakhar1144 replied:
This has been addressed in `mention: Send notifications for @topic wildcard mentions.` commit.

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -210,11 +214,25 @@ def get_recipient_info(
+        if possible_topic_wildcard_mention:
+            # Anyone who sent a message to the topic or reacted to a message on the topic
+            # is considered a possible user to be notified for topic mention.
+            messages = messages_for_topic(recipient.id, stream_topic.topic_name)
+            possibly_topic_wildcard_mentioned_user_ids = (
+                {message.sender_id for message in messages}
+            ) | set(
+                Reaction.objects.filter(message__in=messages).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
```
This will make 2 database queries, one to fetch the sender IDs, and one to fetch the reactor IDs.
Can we avoid that and do a single query on the UserProfile table, maybe using Django's `Q` objects?
Ref: https://docs.djangoproject.com/en/4.2/topics/db/queries/#complex-lookups-with-q-objects

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -265,6 +283,26 @@ def notification_recipients(setting: str) -> Set[int]:
+                if user_id in possibly_topic_wildcard_mentioned_user_ids:
+                    topic_wildcard_mention_user_ids.add(user_id)
+
+        if not possible_wildcard_mention and possible_topic_wildcard_mention:
+            possible_topic_wildcard_mention_user_ids = notification_recipients(
+                "wildcard_mentions_notify"
+            )
```
It would be better to calculate the set of people who have enabled the setting earlier inside an `possible_wildcard_mention OR possible_topic_wildcard_mention` conditional (and call this set simply as `wildcard_mentions_enabled_user_ids` or something), and avoid the super-confusing possible/possibly naming all-together.

prakhar1144 replied:
Avoided the possible/possibly naming and [re-wrote these conditionals](https://github.com/zulip/zulip/pull/25828/commits/514f0def1ad9640fa3ac6baeb8010b314e7bbb17#diff-d803e860e5ff2068cb51aab642015247be71eb4045444a4dcb82e86486f22a4dR276) in a much simpler and better way.

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -265,6 +283,26 @@ def notification_recipients(setting: str) -> Set[int]:
+            possible_topic_wildcard_mention_user_ids = notification_recipients(
+                "wildcard_mentions_notify"
+            )
+            for user_id in possible_topic_wildcard_mention_user_ids:
+                if user_id in possibly_topic_wildcard_mentioned_user_ids:
+                    topic_wildcard_mention_user_ids.add(user_id)
```
This logic is same as that of the one above, right? We should de-duplicate this and put it under a single `if possible_topic_wildcard_mention` conditional.

prakhar1144 replied:
See the [previous comment](https://github.com/zulip/zulip/pull/25828#discussion_r1216089745).

abhijeetbodas2001 reviewed (commented):
I've left some comments on the last commit mostly around naming.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -210,11 +214,25 @@ def get_recipient_info(
+                Reaction.objects.filter(message__in=messages).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
```
> This will make 2 database queries, one to fetch the sender IDs...

`message.sender_id` won't cost an extra db query as the `Message` table already has `sender_id` available.

Ref: https://docs.djangoproject.com/en/4.2/ref/models/fields/#database-representation
> Behind the scenes, Django appends "_id" to the field name to create its database column name.
---
Also, since this query is now getting used in a couple of places ([#20977 > zerver/lib/actions.py](https://github.com/zulip/zulip/commit/1a46df40c33104b685c154a99a57d7bd8dc2c6c0#diff-9a1fd3f8a1bd0ccd686ac2d96d265d3d5dce323f6fc191adac2cb88500af6e1cR6456)),

I'm planning to do a refactoring in a separate commit to have this query inside a lib function, say, `sent_or_reacted_to_messages_user_ids` and use it everywhere.

abhijeetbodas2001 replied:
The reason for that being databases do a lot of work to optimise how the result of a query set is calculated. So telling the database to do the OR would most probably be faster than doing it in memory in Python.

prakhar1144 replied:
This is the updated query:
```python
def participants_for_topic(recipient_id: int, topic_name: str) -> Set[int]:
    """
    Users who either sent or reacted to the messages in the topic.
    The function is expensive for large numbers of messages in the topic.
    """
    messages = Message.objects.filter(recipient_id=recipient_id, subject__iexact=topic_name)
    participants = set(
        UserProfile.objects.filter(
            Q(id__in=Subquery(messages.values("sender_id")))
            | Q(
                id__in=Subquery(
                    Reaction.objects.filter(message__in=messages).values("user_profile_id")
                )
            )
        ).values_list("id", flat=True)
    )
    return participants
```

timabbott replied:
Yeah that looks great! 

The further thing we can do is log the actual database query and then do an `EXPLAIN ANALYZE` on that query to make sure it's using good indexes; but I expect it should be.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -210,11 +214,25 @@ def get_recipient_info(
         # of this function for different message types.
         assert stream_topic is not None
 
+        possibly_topic_wildcard_mentioned_user_ids: AbstractSet[int] = set()
```
`topic_participant_user_ids` sounds better. Changed.

---

> The uncertainty here is not regarding the user IDs set which we want, but rather regarding if there was an actual topic mention in the message, correct?

I was looking at it from a different perspective: since we are uncertain about @topic mention, we are uncertain about the user_ids contained in the list to be notified later in the code flow. (Thinking this way is confusing, actually; I agree.)

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -265,6 +283,26 @@ def notification_recipients(setting: str) -> Set[int]:
             # by default for new users.)
             wildcard_mention_user_ids = notification_recipients("wildcard_mentions_notify")
```
These changes were included as part of the fourth commit.

prakhar1144 commented:
@Ddharmani3 @abhijeetbodas2001

This is ready to review.

Updates in the recent push:
* Addressed the review comments. (Please check the respective comments)
* Added a commit (fourth) to replace [ **wildcard** with **stream_wildcard**](https://github.com/zulip/zulip/pull/25828#discussion_r1210158168)

timabbott commented on zerver/migrations/0453_alter_scheduledmessagenotificationemail_trigger.py:
```diff
@@ -0,0 +1,24 @@
+                ]
+            ),
+        ),
+    ]
```
So this migration is a quirk of Django -- this migration file doesn't actually change the database at all; it just exists to tell the Django migration code that the new choice is a possible valid value in case another migration interacts with it. Rather than create a new migration that doesn't actually change the database, we should edit the last migration touching this field to just have the extra choice included.

prakhar1144 replied:
I think we need to change the value stored in the DB as "wildcard_mentioned" to "stream_wildcard_mentioned" ?

timabbott replied:
Yeah, changing existing values in the database sounds correct -- but I believe this migration file just changes the set of values that Django's Python implementation thinks are valid values, without changing existing values in the database at all.

abhijeetbodas2001 replied:
We'll need to do two things here:
1. Edit the possible choices to what we want. Edit 0331_scheduledmessagenotificationemail.py as well. Reasoning: https://chat.zulip.org/#narrow/dm/16839,22341-dm/near/1592884
2. To update the existing values in the DB, create a custom migration. See https://github.com/zulip/zulip/pull/24354/commits/0cfa9c96f58841f531987a9422f118ea6fed594d for an example.

abhijeetbodas2001 replied:
@timabbott does that sound right?

prakhar1144 replied:
In the first commit:
1.[ Edit 0031_scheduledmessagenotificationemail.py](https://github.com/zulip/zulip/pull/25828/commits/b4aa15502d467199e2ee706def223dc21b988abb#diff-932e136b53606abc4de3a5f5fded9fb351850927c044dc24812c98eab2773843)
2. [Update the existing values in DB](https://github.com/zulip/zulip/pull/25828/commits/b4aa15502d467199e2ee706def223dc21b988abb#diff-9f450075da1f52de9653a804b2dfb3d74ae674f68df480c98e897b1d7d036e6a)

timabbott replied:
Yeah that's correct.

timabbott commented on zerver/models.py:
```diff
@@ -4291,7 +4291,7 @@ class ScheduledMessageNotificationEmail(models.Model):
         (NotificationTriggers.PRIVATE_MESSAGE, "Private message"),
         (NotificationTriggers.MENTION, "Mention"),
-        (NotificationTriggers.WILDCARD_MENTION, "Wildcard mention"),
+        (NotificationTriggers.STREAM_WILDCARD_MENTION, "Stream wildcard mention"),
```
This string doesn't appear in the UI, right?

prakhar1144 replied:
Right, It doesn't appear in the UI.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1193,7 +1201,7 @@ def process_message_update_event(
                 stream_push_user_ids=stream_push_user_ids,
                 stream_email_user_ids=stream_email_user_ids,
-                wildcard_mention_user_ids=wildcard_mention_user_ids,
+                stream_wildcard_mention_user_ids=stream_wildcard_mention_user_ids,
```
This module needs to support the old format, because there can be events containing a `wildcard_mention_user_ids` field in the Tornado queues at the time the Zulip server is upgraded, and we need those to not break. Usually it's pretty simple -- if a field by that name exists, replace it with a `stream_wildcard_mention_user_ids` field. 

Search for `TODO/compatibility` in this file for examples of how to do this.

There may be a similar issue for the output of `maybe_enqueue_notifications`, too -- we should either avoid changing the format of what is enqueued by that function, or add similar compatibility code.

prakhar1144 replied:
As suggested in [#25828 (comment)](https://github.com/zulip/zulip/pull/25828#issuecomment-1581675528) -- addressed this in #25978

timabbott commented:
I merged the first 3 commits as the series ending with 79e5d32ef6bfa9ad5e95f503118e56931ccb27b9. The 4th commit might be worth extracting to a separate PR to make it easier to verify we're handling https://github.com/zulip/zulip/pull/25828#discussion_r1222293742 correctly; I think the rename is a good change but we want to be really careful, because bugs of that form can require manual intervention to handle in production.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -210,11 +214,23 @@ def get_recipient_info(
+                Reaction.objects.filter(message__in=messages).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
```
This should probably be turned into a `zerver.lib.topic` method `participants_for_topic` -- it's a clear enough definition that we may find we want to use it for something else later, and I think it'll help readability. 

We should also document that function as potentially expensive -- the `Message` part is pretty cheap, but the `Reaction` part may or may not be if there's a lot of messages in the topic. And add a comment here that it's OK to do this potentially expensive thing since we only do it when `@topic` appears in the message body.

(It's also the sort of thing that could be optimized to a single query if profiling suggested that's useful).

prakhar1144 replied:
[#25828 Comment](https://github.com/zulip/zulip/pull/25828#discussion_r1240792510)

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -257,13 +273,21 @@ def notification_recipients(setting: str) -> Set[int]:
+        if possible_topic_wildcard_mention:
+            for user_id in wildcard_mentions_notify_user_ids:
+                if user_id in topic_participant_user_ids:
+                    topic_wildcard_mention_user_ids.add(user_id)
```
I feel like this probably should be written using the Python set intersection syntax that probably takes care of doing the loop order starting from the smallest set for us, but if we're doing this loop structure you definitely want the other loop order, because `wildcard_mentions_notify_user_ids` is probably everyone subscribed to the stream, whereas `topic_wildcard_mention_user_ids` is typically going to be the small subset who've participated in this topic.

prakhar1144 replied:
>  this probably should be written using the Python set intersection syntax that probably takes care of doing the loop order starting from the smallest set for us

Yes, verified with [the docs](https://wiki.python.org/moin/TimeComplexity).

* Average Time Complexity for `intersection` is `O(min(len(setA), len(setB)))`
* Worst Time Complexity is `O(len(setA)*len(setB))` -- I think this would be the case only when accessing element from map is `O(n)`, which is rare.

In short, we have,
`topic_wildcard_mention_user_ids` = `topic_participant_user_ids.intersection(wildcard_mentions_notify_user_ids)`.

timabbott commented on templates/zerver/emails/missed_message.html:
```diff
@@ -29,6 +29,8 @@
     {% elif mentioned_user_group_name %}
     {% trans %}You are receiving this because @{{ mentioned_user_group_name }} was mentioned.{% endtrans %}<br />
+    {% elif topic_wildcard_mentioned %}
+    {% trans %}You are receiving this because everyone was mentioned in #{{ stream_name }} > {{ topic_name }}.{% endtrans %}<br />
```
This should be discussed in #design, but I was thinking of something more like "because @topic was mentioned in #{{ stream_name }} > {{ topic_name }}.` or "because all participants were mentioned in #{{ stream_name }} > {{ topic_name }}.`

prakhar1144 replied:
I made the changes after [discussion in CZO.](https://chat.zulip.org/#narrow/stream/101-design/topic/notification.20emails/near/1588768)

> You are receiving this because all topic participants were mentioned in #{{ stream_name }} > {{ topic_name }}.

timabbott commented on zerver/tests/test_email_notifications.py:
```diff
@@ -1829,6 +1916,20 @@ def test_long_term_idle_user_missed_message(self) -> None:
                 [{"message_id": personal_message_id, "trigger": "private_message"}],
             )
 
+        # Topic Wildcard mention should NOT soft reactivate the user
```
I think actually it makes sense to trigger soft-reactivating topic participants -- the rationale being that generally mentioning a topic is going to reach a small set of users who are probably already active, and anyone who is not already active has a decent change of being reactivated by the mention.

prakhar1144 replied:
Makes sense.

I added a separate commit for this. (Last commit)

timabbott commented:
This is great progress; I finished a read through the branch and commented on most of my concerns.

prakhar1144 commented:
**Updates in the recent push:**
* Addressed the review comments (Please check the respective review comments)
* Opened #25978 
* Added missing tests in [`test_push_notifications.py`](https://github.com/zulip/zulip/pull/25828/commits/21651491e9cd293aea179516c56ad1ede3f48519#diff-48b4e166fb7d503e72d556c486f4b1cc2cbe3c56c92942786a0bd29072a20ca5)
  * Alert in the case of apns:
  
    ```python
    "alert": {
        "title": "#Verona > Test topic",
        "subtitle": "King Hamlet mentioned all the topic participants:",
        "body": message.content,
     },
    ```
  * gcm alert
     ```
     King Hamlet mentioned all the topic participants in #Verona > Test topic
     ```

**We're left with two things to discuss:**
* Django migration related doubt. -- [comment](https://github.com/zulip/zulip/pull/25828#discussion_r1223845341)
* How to handle topic participants calculation.
  * First one is the approach I've currently used: First commit creates a lib function which gets reused to calculate topic participants. --  [comment](https://github.com/zulip/zulip/pull/25828#discussion_r1222296879)
  * Second approach -- [Abhijeet's suggestion to do this in one query](https://github.com/zulip/zulip/pull/25828#discussion_r1224072475) -- sounds good to me. Let's have the final decision, in that case I'll drop the first commit.

prakhar1144 commented:
Implemented:
* Email and push notifications are sent to the topic participants for the @topic wildcard mention.

Updates in the recent push:
* After #25640 got merged. We have `followed_topic_stream_wildcard_mention` and `followed_topic_topic_wildcard_mention` -- parallel to `topic_wildcard_mention` and `stream_wildcard_mention`.

  To explain a bit:
  * `followed_topic_stream_wildcard_mention_user_ids` stores the ids of the users on the stream who follow the topic and have enabled the wildcard mention notifications for followed topics.
  * `followed_topic_topic_wildcard_mention_user_ids` stores the ids of the topic participants who follow the topic and have enabled the wildcard mention notifications for followed topics.

  Made all the changes related to this.
 
 * Addressed the review comments -- please check the respective review comments.
 * 2nd, 3rd, 4th and 5th are new commits which fixes things we missed in #25640 
---
Awaiting feedback:

abhijeetbodas2001 commented:
Let's split this PR into smaller ones here on (since this is getting a bit unwieldy); it would be great to cherry-pick the bug fix commits here onto #25978, along with the big rename commit.

abhijeetbodas2001 commented:
> Should we introduce a new UserMessage flag named topic_wildcard_mentioned and rename the existing flag to stream_wildcard_mentioned

I think that is best avoided, but touching the `UserMessage` table has a lot of performance implications, so please open a CZO topic to discuss this.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -194,14 +194,14 @@ def get_recipient_info(
     followed_topic_push_user_ids: Set[int] = set()
     followed_topic_email_user_ids: Set[int] = set()
-    followed_topic_wildcard_mention_user_ids: Set[int] = set()
+    followed_topic_stream_wildcard_mention_user_ids: Set[int] = set()
```
[CZO Discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/.40topic.20mention/near/1600048)

prakhar1144 commented on zerver/tornado/event_queue.py:
```diff
@@ -1149,11 +1184,22 @@ def process_message_update_event(
     stream_email_user_ids = set(event_template.pop("stream_email_user_ids", []))
-    wildcard_mention_user_ids = set(event_template.pop("wildcard_mention_user_ids", []))
+
+    # TODO/compatibility: Translation code for the rename of
+    # `wildcard_mention_user_ids` to `stream_wildcard_mention_user_ids`.
+    # Remove this when one can no longer directly upgrade from 7.x to main.
```
[As discussed earlier](https://chat.zulip.org/#narrow/stream/378-api-design/topic/allowed.20visibility.20policies/near/1580224):
>   I don't think it's necessarily a problem to have the API allow you to follow topics -- it may be useful for testing parts of the feature on [chat.zulip.org](http://chat.zulip.org/) before we're ready to turn it on for everyone, and I don't think we intend to have intermediate states where it is a security problem that you were allowed to follow a topic.

> We just don't guarantee to users that it'll be a good experience to do so.

So, should we care about providing a good experience?  Our docs don't have "Follow" as a visibility policy as of now.

abhijeetbodas2001 replied:
Makes sense. We just need to make sure the server doesn't throw errors because of this (`.get` without a default value, or `.pop` can throw errors if the key isn't present in the dict, so we need to make sure isn't any such code).
Users not receiving notifications for followed topics is fine, but runtime errors aren't

prakhar1144 replied:
We have default value available.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -292,14 +310,31 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
             followed_topic_stream_wildcard_mention_user_ids = (
-                followed_topic_notification_recipients("wildcard_mentions_notify")
+                followed_topic_wildcard_mentions_notify_user_ids
+            )
```
Linters adds it.

abhijeetbodas2001 replied:
Ah I see :)

prakhar1144 commented on zerver/tests/test_email_notifications.py:
```diff
@@ -1101,21 +1101,56 @@ def test_user_group_over_stream_wildcard_mention_priority(self) -> None:
+        for text in expected_email_include:
+            self.assertIn(text, self.normalize_string(mail.outbox[0].body))
+
+    def test_stream_wildcard_over_followed_topic_mention_priority(self) -> None:
+        hamlet = self.example_user("hamlet")
+        othello = self.example_user("othello")
+
+        followed_topic_mentioned_message_id = self.send_stream_message(othello, "Denmark", "1")
+        stream_wildcard_mentioned_message_id = self.send_stream_message(
+            othello, "Denmark", "@**all**"
+        )
+
+        handle_missedmessage_emails(
+            hamlet.id,
+            [
+                {
+                    "message_id": followed_topic_mentioned_message_id,
+                    "trigger": "followed_topic_email_notify",
```
* We still have n assertions for n possible triggers
* I think it's better to test by sending separate messages having different elements (personal mention / wildcards), instead of a single message having all those elements. The reason being: it's more practical that a user has a few different missed messages having different elements and the email text priority is decided based on triggers of all those messages. a single message having all those triggers doesn't seems a practical case (although it will also verify the code logic).
* Having separate tests seems more readable to me.

abhijeetbodas2001 replied:
Cool, makes sense to me :)

abhijeetbodas2001 replied:
It seems the tests are patterned as `assert(P1>P2)`, `assert(P2>P3)`, ... where `Px` is priority `x`. So `n-1` tests in total :)
Might be helpful to add a comment here explaining the testing pattern, but not a blocker.
Also, I noticed that we currently have an existing test called `test_wildcard_over_stream_mention_priority`, the name of which is kinda misleading. `test_wildcard_mention_over_stream_message_priority` might be better?

prakhar1144 replied:
[#25978 (comment)](https://github.com/zulip/zulip/pull/25978#discussion_r1247709379)

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -257,7 +257,7 @@ def get_mentions_for_message_updates(message_id: int) -> Set[int]:
     rendering_result: MessageRenderingResult, ums: Iterable[UserMessage]
 ) -> None:
-    stream_wildcard = rendering_result.mentions_stream_wildcard
+    wildcard_mentioned = rendering_result.has_wildcard_mention()
```
Makes sense. Changes done.

prakhar1144 commented on zerver/tests/test_users.py:
```diff
@@ -1820,6 +1823,55 @@ def test_stream_recipient_info(self) -> None:
         self.assertEqual(info.stream_wildcard_mention_user_ids, {hamlet.id, othello.id})
 
+        hamlet.wildcard_mentions_notify = True
+        hamlet.save()
```
Used `do_change_user_settings` here.

I'll create a separate PR to follow the suggested pattern. (This PR is already too large :))

abhijeetbodas2001 replied:
Cool!

prakhar1144 replied:
> I'll create a separate PR to follow the suggested pattern. (This PR is already too large :))

Opened #26404

prakhar1144 commented on zerver/lib/push_notifications.py:
```diff
@@ -672,6 +672,11 @@ def get_gcm_alert(
+        message.is_stream_message()
+        and trigger == NotificationTriggers.FOLLOWED_TOPIC_STREAM_WILDCARD_MENTION
+    ):
+        return "TODO"
```
[CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/notification.20emails/near/1598232)

prakhar1144 commented on zerver/migrations/0454_backfill_scheduledmessagenotificationemail_trigger.py:
```diff
@@ -0,0 +1,24 @@
+
+    operations = [
+        migrations.RunSQL(
+            """
+            UPDATE zerver_scheduledmessagenotificationemail
+            SET trigger = 'stream_wildcard_mentioned'
+            WHERE trigger = 'wildcard_mentioned';
+            """,
+            reverse_sql="""
+            UPDATE zerver_scheduledmessagenotificationemail
+            SET trigger = 'wildcard_mentioned'
+            WHERE trigger = 'stream_wildcard_mentioned';
+            """,
```
Added migration for `follow_topic` equivalent.

prakhar1144 commented on zerver/lib/email_notifications.py:
```diff
@@ -525,7 +527,9 @@ def do_send_missedmessage_events_reply_in_zulip(
-                    if m["trigger"] == "mentioned" or m["trigger"] == "stream_wildcard_mentioned"
+                    if m["trigger"] == "mentioned"
+                    or m["trigger"] == "stream_wildcard_mentioned"
+                    or m["trigger"] == "followed_topic_stream_wildcard_mentioned"
```
**mention_count context variable:**

`mention_count` is stale variable. Introduced in bc861b4441f129cfb1bb6dc21243d8424fb0dac1 then never got used. -- Removed that in the first commit.

**Regarding the senders list calculation here:**

* This is also stale and will be removed.

* I removed it locally but one test case failed. On investigating i found a couple of bugs which i have [explained in CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/send.20missed.20message.20emails.20as.20user/near/1601380).

* Once those bugs are fixed and test updated, we can safely remove this code block.

* Ideally, this should have been removed in c4d28777ec6bf62cf0e0f9217b3ba09597161afd -- as we no longer use `sender_str` as subject for mentions.

abhijeetbodas2001 replied:
Makes sense. Great to have caught these bugs :)

prakhar1144 commented on zerver/tests/test_markdown.py:
```diff
@@ -1822,6 +1822,7 @@ def test_mention_wildcard(self) -> None:
+        self.assertFalse(rendering_result.mentions_topic_wildcard)
         self.assertTrue(rendering_result.mentions_stream_wildcard)
 
     def test_mention_everyone(self) -> None:
```
[#25978 (Comment)](https://github.com/zulip/zulip/pull/25978#discussion_r1247685183)

prakhar1144 commented on zerver/tests/test_markdown.py:
```diff
@@ -1822,6 +1822,7 @@ def test_mention_wildcard(self) -> None:
             '<p><span class="user-mention" data-user-id="*">@all</span> test</p>',
         )
+        self.assertFalse(rendering_result.mentions_topic_wildcard)
         self.assertTrue(rendering_result.mentions_stream_wildcard)
```
[#25978 (Comment)](https://github.com/zulip/zulip/pull/25978#discussion_r1247685183)

timabbott commented:
> Should we introduce a new UserMessage flag named topic_wildcard_mentioned and rename the existing flag to stream_wildcard_mentioned

Yeah, I think I also prefer to not change that structure for now. I don't think it's necessary for the logic required for this feature, and if we change it, we'll probably need to do a bigger overhaul that also makes group mentions cleaner.

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -293,14 +311,31 @@ def followed_topic_notification_recipients(setting: str) -> Set[int]:
         )
         followed_topic_push_user_ids = followed_topic_notification_recipients("push_notifications")
 
+        if possible_stream_wildcard_mention or possible_topic_wildcard_mention:
+            # We calculate `wildcard_mentions_notify_user_ids` and `followed_topic_wildcard_mentions_notify_user_ids`
+            # only if there's a possible stream or topic wildcard mention in the message.
+            # This is important so as to avoid unnecessarily sending huge user ID lists with
+            # thousands of elements to the event queue (which can happen because these settings
+            # are `True` by default for new users.)
+            wildcard_mentions_notify_user_ids = notification_recipients("wildcard_mentions_notify")
+            followed_topic_wildcard_mentions_notify_user_ids = (
+                followed_topic_notification_recipients("wildcard_mentions_notify")
+            )
+
         if possible_stream_wildcard_mention:
-            # We calculate `stream_wildcard_mention_user_ids` and `followed_topic_wildcard_mention_user_ids`
-            # only if there's a possible stream wildcard mention in the message. This is important so as
-            # to avoid unnecessarily sending huge user ID lists with thousands of elements to the
-            # event queue (which can happen because these settings are `True` by default for new users.)
-            stream_wildcard_mention_user_ids = notification_recipients("wildcard_mentions_notify")
+            stream_wildcard_mention_user_ids = wildcard_mentions_notify_user_ids
             stream_wildcard_mention_in_followed_topic_user_ids = (
-                followed_topic_notification_recipients("wildcard_mentions_notify")
+                followed_topic_wildcard_mentions_notify_user_ids
```
This code looks much cleaner than the first version!

One question I have is, should we be doing a deep/shallow copy of the common variables (`wildcard_mentions_notify_user_ids` and `followed_topic_wildcard_mentions_notify_user_ids`)? I don't remember how Python does copying of variables etc, but I do remember than its a bit different than other languages like C++/Java.

If not that, we should at least make sure that we have a test simulating a scenario of a message containing both stream and topic mentions, and then verify that our user IDs sets are the same as what we expect.

abhijeetbodas2001 replied:
Oh I see we have a test for that, great! :)

prakhar1144 replied:
Had a discussion around this with Abhijeet on CZO (DM):

Here it makes a shallow copy. But even if the sets are changed at a later stage, it will not affect the behavior as we are no longer going to reuse the `followed_topic_wildcard_mentions_notify_user_ids` or `wildcard_mentions_notify_user_ids`.

And we already have tests to cover the expected behavior. So, we're good to go.

abhijeetbodas2001 reviewed (commented):
I reviewed all but the last 2 commits, and they mostly look good to me (some minor comments below).

> tests: Update tests to cover the corner case properly.

This commit message is kinda vague; we should mention which area in the tests this is touching.

I think the group of first 6 commits would be a good prefix to split into a new PR so that it can be merged off. The notifications commit is pretty large and would be easier to review independently :)

prakhar1144 commented on zerver/tests/test_email_notifications.py:
```diff
@@ -1971,7 +1971,6 @@ def test_long_term_idle_user_missed_message(self) -> None:
 
         # Stream Wildcard mention should NOT soft reactivate the user
         with self.soft_deactivate_and_check_long_term_idle(hamlet, expected=True):
-            # Soft reactivate the user by sending a personal message
             mention = "@**all**"
```
Done

prakhar1144 commented on zerver/tests/test_email_notifications.py:
```diff
@@ -1955,6 +1964,17 @@ def test_long_term_idle_user_missed_message(self) -> None:
+        )
+        do_change_user_setting(hamlet, "wildcard_mentions_notify", False, acting_user=None)
+
         # Stream wildcard mention in followed topic should NOT soft reactivate the user
```
Done

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -731,6 +743,15 @@ def create_user_messages(
         if user_profile_id in ids_with_alert_words:
             flags |= UserMessage.flags.has_alert_word
+        if (
+            not rendering_result.mentions_stream_wildcard
```
This `not` is a noop, because that flag will already be there, right? Might be cleaner to not have it.

prakhar1144 replied:
Suppose a message has both `stream` and `topic` wildcard mentions.

We already set the `wildcard_mentioned` flag for all the `um_eligible_user_ids` earlier in this function due to the stream wildcard mention. 
```python
    base_flags = 0
    if rendering_result.mentions_stream_wildcard:
        base_flags |= UserMessage.flags.wildcard_mentioned
```
So using this `not`, we can skip the following operation in the `if` conditional
```python
user_profile_id in topic_wildcard_mention_user_ids.union(
    topic_wildcard_mention_in_followed_topic_user_ids
)
```

In simple words, if the message has both the stream and topic wildcard mention we can skip the checks related to topic wildcard (using this not) as we already have set the `wildcard_mentioned` flag for everyone. No need to check and set it again for individual users.

prakhar1144 replied:
Suppose a message has both `stream` and `topic` wildcard mentions.

We already set the `wildcard_mentioned` flag for all the `um_eligible_user_ids` earlier in this function due to the stream wildcard mention. 
```python
    base_flags = 0
    if rendering_result.mentions_stream_wildcard:
        base_flags |= UserMessage.flags.wildcard_mentioned
```

So using this `not`, we can skip the following operation in the if conditional:

```python
user_profile_id in topic_wildcard_mention_user_ids.union(
   topic_wildcard_mention_in_followed_topic_user_ids
)
```

In simple words, if the message has both the stream and topic wildcard mention, we can skip the checks related to topic wildcard (using this `not`), as we already have set the `wildcard_mentioned` flag for everyone. There is no need to check and set it again for individual users.

prakhar1144 replied:
One improvement I made:

Instead of performing the union operation within a loop, I extracted that as a separate variable named `all_topic_wildcard_mention_user_ids`

abhijeetbodas2001 replied:
> we can skip the checks related to topic wildcard (using this not)

That makes sense, but I'm not sure if this optimization would make that much of a difference, since all the data here is in-memory?

prakhar1144 replied:
Hm, right. Earlier the union operation was there in the if statement itself. Since I've extracted it out, we can skip this `not` check.

timabbott commented on zerver/lib/email_notifications.py:
```diff
@@ -456,12 +456,17 @@ def do_send_missedmessage_events_reply_in_zulip(
+        or "topic_wildcard_mentioned" in unique_triggers
         or "stream_wildcard_mentioned" in unique_triggers
+        or "topic_wildcard_mentioned_in_followed_topic" in unique_triggers
         or "stream_wildcard_mentioned_in_followed_topic" in unique_triggers,
```
This first block might be better extracted as a separate variable.

prakhar1144 replied:
Done

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -842,8 +849,14 @@ def get_apns_alert_subtitle(
         return _("TODO")
+    elif trigger == NotificationTriggers.TOPIC_WILDCARD_MENTION:
+        return _("{full_name} mentioned all the topic participants:").format(
+            full_name=message.sender.full_name
```
maybe "mentioned all topic participants:" would be better -- I don't think we need the " the " here.

prakhar1144 replied:
Done.

Also removed "the" at line 686 in this file.

prakhar1144 replied:
Done.

Also removed "the" at line 686 in this file.

timabbott commented:
I reviewed everything except I didn't do a careful read on the tests in the last couple commits; @abhijeetbodas2001 it'd be great if you can put some extra attention into those. Meanwhile, @prakhar1144 I think you can start working on draft /help/ documentation for the feature -- it seems like we're getting close to integrating the feature.

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments

Blockers:
* Push notification texts  [#25828 (comment)](https://github.com/zulip/zulip/pull/25828/commits/6f7a2e659aba11a01823d4cbbbb4973d841a63c3#r1246961247) -- Maybe we can update this with [#25978 (comment)](https://github.com/zulip/zulip/pull/25978#issuecomment-1619492173) in a separate PR.
* [#25828 (comment)](https://github.com/zulip/zulip/pull/25828#issuecomment-1624416034) :
> I reviewed everything except I didn't do a careful read on the tests in the last couple commits; @abhijeetbodas2001 it'd be great if you can put some extra attention into those.

prakhar1144 commented:
Updates in the recent push:
* The recent push includes changes explained in [#26195 (comment)](https://github.com/zulip/zulip/pull/26195#issuecomment-1630888200)
* Updates for this PR is summarized in the [above comment](https://github.com/zulip/zulip/pull/25828#issuecomment-1628268644)

abhijeetbodas2001 commented on zerver/tests/test_soft_deactivation.py:
```diff
@@ -770,6 +772,17 @@ def assert_stream_message_not_sent_to_idle_user(
 
+        # Test UserMessage row is created while user is deactivated if
+        # there is a topic wildcard mention i.e. @topic
+        do_soft_activate_users([long_term_idle_user])
+        self.send_stream_message(long_term_idle_user, stream_name, "Hi", topic_name)
+        topic_participant_user_ids = {long_term_idle_user.id}
+
+        assert_stream_message_sent_to_idle_user(
+            "Test @**topic** mention", topic_participant_user_ids=topic_participant_user_ids
+        )
+        do_soft_deactivate_users([long_term_idle_user])
```
Should we be first deactivating, then testing here? We need to check that message is created while the user is soft-deactivated, correct?

abhijeetbodas2001 reviewed (commented):
I read through the tests, mostly look fine to me :+1:

prakhar1144 commented on templates/zerver/emails/missed_message.html:
```diff
@@ -29,8 +29,12 @@
     {% trans %}You are receiving this because you were personally mentioned.{% endtrans %}<br />
     {% elif mentioned_user_group_name %}
     {% trans %}You are receiving this because @{{ mentioned_user_group_name }} was mentioned.{% endtrans %}<br />
+    {% elif topic_wildcard_mentioned_in_followed_topic %}
+    {% trans %}You are receiving this because all topic participants were mentioned in #{{ stream_name }} > {{ topic_name }}.{% endtrans %}<br />
     {% elif stream_wildcard_mentioned_in_followed_topic %}
     {% trans %}You are receiving this because you have wildcard mention notifications enabled for topics you follow.{% endtrans %}<br />
+    {% elif topic_wildcard_mentioned %}
+    {% trans %}You are receiving this because all topic participants were mentioned in #{{ stream_name }} > {{ topic_name }}.{% endtrans %}<br />
```
Yes -- [CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/notification.20emails/near/1603388)
> OK, I think that plan is reasonable and certainly simplest -- we can always change that detail if we find we get feedback that suggests a different model in this rare corner case.

prakhar1144 commented on zerver/tests/test_email_notifications.py:
```diff
@@ -741,6 +783,47 @@ def _extra_context_in_missed_stream_messages_stream_wildcard_mention_in_followed
+            send_as_user,
+            show_message_content=show_message_content,
+            verify_body_does_not_include=verify_body_does_not_include,
+            trigger="topic_wildcard_mentioned",
```
Included in my TODO list. -- stray trigger strings have been used in a lot of places in this file and a few others. I'll change all of them in a separate mini PR.

prakhar1144 commented:
I have addressed the review comments.

[Summary](https://github.com/zulip/zulip/pull/25828#issuecomment-1630896318)

prakhar1144 commented:
@zulipbot add "integration review"

timabbott commented on zerver/lib/message.py:
```diff
@@ -168,7 +168,10 @@ class SendMessageRequest:
     service_bot_tuples: List[Tuple[int, int]]
     all_bot_user_ids: Set[int]
+    topic_wildcard_mention_user_ids: Set[int]
     stream_wildcard_mention_user_ids: Set[int]
```
Does `topic_wildcard_mention_user_ids` include `topic_wildcard_mention_in_followed_topic_user_ids`? It might be really good to have a comment here making that extremely clear.

prakhar1144 replied:
Comment added.

timabbott commented on zerver/lib/message.py:
```diff
@@ -168,7 +168,10 @@ class SendMessageRequest:
     stream_wildcard_mention_user_ids: Set[int]
+    # IDs of users who have followed the topic the message (having topic wildcard) is being sent to, and have the followed topic wildcard mentions notify setting ON.
+    topic_wildcard_mention_in_followed_topic_user_ids: Set[int]
     # IDs of users who have followed the topic the message (having stream wildcard) is being sent to, and have the followed topic wildcard mentions notify setting ON.
```
These comments should be line wrapped to 80 characters.

prakhar1144 replied:
Done

timabbott commented:
I posted a couple small comments; otherwise lgtm. It appears that the conclusion of the discussion of the GCM/APNS alerts is that we can probably remove that code as not doing anything, so the TODO strings are maybe fine for this PR and we can do that cleanup in a follow-up.

prakhar1144 commented:
@timabbott Addressed the review comments.

timabbott commented:
Merged, thanks @prakhar1144! Excited to have another of these big PRs integrated :)
